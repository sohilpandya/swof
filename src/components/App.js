import React, { Component } from 'react';
import update from 'immutability-helper';
import Clock from './Clock';

import getDay from '../utils/getDay';
import getAvailableEngineers from '../utils/getAvailableEngineers';
import initialState from '../data';
import { log } from 'util';


class App extends Component {

  constructor(props) {
    super(props)

    this.state = initialState;
  }

  render() {
    return (
      <div className="vh-100 bg-primary">
        <div className="fl w-100 vh-25">
          <p className="tc f1 b secondary mb0">Wheel Of Fate</p>
          <Clock />
        </div>

        <div className="fl w-100 vh-50 secondary">
          {
            this.state.TodaysEngineers.length === 0 &&
            <div>
              <div className="w-50 tc bg-secondary-50 primary pa2 b pointer m-auto br3 dim"
                    onClick={() => { this.generateShift() }}>
                      Generate Shift
                      <div className="f7"> please click here to generate the shifts for today </div>
              </div>
            </div>
          }
          <div className="fl w-50 vh-50 tc">
            {
              this.state.TodaysEngineers.length > 0 &&
              <div>
                { new Date().getHours() <= 13 && new Date().getHours >= 8 ?
                  <div className="loading"></div> : <div className="empty-loading"></div>
                }
                <h2> Morning Shift </h2>
                <h1> {this.state.TodaysEngineers[0].name} </h1>
              </div>
            }
          </div>
          <div className="fl w-50 vh-50 tc">
            {
              this.state.TodaysEngineers.length > 0 &&
              <div>
                { new Date().getHours() > 14 ?
                  <div className="loading"></div> : <div className="empty-loading"></div>
                }
                <h2> Afternoon Shift </h2>
                <h1> {this.state.TodaysEngineers[1].name} </h1>
              </div>
            }
          </div>
        </div>

        <div className="fl w-100 vh-25 secondary">
          <div className="flex justify-around flex-wrap h-100 b">
            {this.getEmployeeNames()}
          </div>
        </div>

      </div>
    );
  }

  componentDidMount() {
    this.setState({
      ...this.state,
      day: getDay()
    })
  }

  getEmployeeNames = () => {
    return this.state.employees.map((x,i) => {
      return (
        <div className="flex flex-column flex-grow-1 bt b--secondary items-center justify-center" key={i}>
            <div>{x.name}</div>
            <div> No. Shifts {x.totalShifts}</div>
        </div>
      )
    })
  }

  generateShift = () => {

    let newEmployeesState = this.state.employees;
    let newTodaysEngineers = [];

    for (var i = 0; i < 2; i++) {
      let availableEngineers = getAvailableEngineers(newEmployeesState);
      let rng = Math.floor(Math.random() * (availableEngineers.length))
      let engineer = availableEngineers[rng];
      let engineerIndex = newEmployeesState.findIndex((x) => x.name === engineer.name)
      let newEngineerDetails = update(engineer, {
        $set: {
          name: engineer.name,
          lastWorked: this.state.day,
          workingToday: true,
          totalShifts: engineer.totalShifts + 1
        }
      })

      newTodaysEngineers.push(newEngineerDetails);
      newEmployeesState = update(newEmployeesState, {
        $splice: [
          [engineerIndex, 1, newEngineerDetails]
        ]
      })

    }

    this.setState({
      ...this.state,
      TodaysEngineers: newTodaysEngineers,
      employees: newEmployeesState
    }, () => { console.log('state has been updated') })

  }
}

export default App;
