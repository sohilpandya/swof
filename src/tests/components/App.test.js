import React from 'react';
import ReactDOM from 'react-dom';
import App from '../../components/App';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
});


test('App Component shows intial information when firt loaded', () => {

  // 1. Initial render of the component
  const component = renderer.create(
    <App />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // // 2. Manually Trigger getEmployeeName
  // console.log(tree);

  // tree.props.getEmployeeName();
  //   // rerenders the component
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
});
