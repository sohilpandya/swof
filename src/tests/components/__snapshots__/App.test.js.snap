// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component shows intial information when firt loaded 1`] = `
<div
  className="vh-100 bg-primary"
>
  <div
    className="fl w-100 vh-25"
  >
    <p
      className="tc f2 secondary"
    >
      Wheel Of Fate
    </p>
  </div>
  <div
    className="fl w-100 vh-50 secondary"
  >
    <button
      onClick={[Function]}
    >
      Wheel of Fortune
    </button>
    <div
      className="fl w-50 vh-50"
    />
    <div
      className="fl w-50 vh-50"
    />
  </div>
  <div
    className="fl w-100 vh-25"
  >
    <div
      className="flex justify-around flex-wrap h-100 b"
    >
      <div
        className="flex flex-grow-1 ba b--blue items-center justify-center"
      >
        Sohil
      </div>
      <div
        className="flex flex-grow-1 ba b--blue items-center justify-center"
      >
        Alex
      </div>
      <div
        className="flex flex-grow-1 ba b--blue items-center justify-center"
      >
        Andrew
      </div>
      <div
        className="flex flex-grow-1 ba b--blue items-center justify-center"
      >
        Joe
      </div>
      <div
        className="flex flex-grow-1 ba b--blue items-center justify-center"
      >
        Rory
      </div>
      <div
        className="flex flex-grow-1 ba b--blue items-center justify-center"
      >
        Mavis
      </div>
      <div
        className="flex flex-grow-1 ba b--blue items-center justify-center"
      >
        June
      </div>
      <div
        className="flex flex-grow-1 ba b--blue items-center justify-center"
      >
        Dan
      </div>
      <div
        className="flex flex-grow-1 ba b--blue items-center justify-center"
      >
        Rose
      </div>
      <div
        className="flex flex-grow-1 ba b--blue items-center justify-center"
      >
        Cleo
      </div>
    </div>
  </div>
</div>
`;
